[Ryan Low] iP

Ryan-L98 opened the pull request:
![image](https://i.dailymail.co.uk/1s/2021/07/17/23/45569135-9798849-image-a-44_1626560944347.jpg)
# Bob
A task tracker to help you remember what when and if you already completed a task.
> The best task tacker I have ever used
>> it has made my life so much worse - me

### Types of tasks Bob can currently track
1. To do 
2. Events
3. Deadlines

### Other things Bob can do
- `List` out your tasks
- `Find` tasks that you have added
- `Mark` tasks as done
- `Delete` your tasks

### Demo
I want to add a new deadline using Bob.
```
██╗░░██╗██╗  ██╗██╗███╗░░░███╗  ██████╗░░█████╗░██████╗░
██║░░██║██║  ██║╚█║████╗░████║  ██╔══██╗██╔══██╗██╔══██╗
███████║██║  ██║░╚╝██╔████╔██║  ██████╦╝██║░░██║██████╦╝
██╔══██║██║  ██║░░░██║╚██╔╝██║  ██╔══██╗██║░░██║██╔══██╗
██║░░██║██║  ██║░░░██║░╚═╝░██║  ██████╦╝╚█████╔╝██████╦╝
╚═╝░░╚═╝╚═╝  ╚═╝░░░╚═╝░░░░░╚═╝  ╚═════╝░░╚════╝░╚═════╝░

====================================================================
Bob: What tasks are you adding to the list today? 	¯\(°_o)/¯
	No tasks saved previously~
====================================================================
You: deadline Week 4 iP increments /by 2022-02-03T23:59
====================================================================
Bob: I have added Week 4 iP increments to your tasks! You have 1 tasks now ._.)/\(._.
====================================================================
You: list
====================================================================
Bob: Okay, these are your tasks
	1. [D] [ ] Week 4 iP increments (by: 03 Feb 2022 23:59)
      (づ｡◕‿‿◕｡)づ
====================================================================
```

## Random code block from Bob
```Java
public void initializeBob() {
   store = new Storage("Bob.txt");
   tasks = new TaskList(store.getSavedStore());
   ui = new Ui();
   greet();
   activeListener();
}
```

***Please*** *download the first version of Bob from any of these links*.
| Click the link below | or u can also use this link |
| --- | ----------- |
| [link](https://github.com/Ryan-L98/ip/releases/tag/A-Jar) | [link again](https://github.com/Ryan-L98/ip/releases/tag/A-Jar) |

### Everything included in this README 😜 😋 😝
- [x] a heading
- [x] a bullet list
- [x] a numbered list
- [x] a fenced code block (with syntax highlighting)
- [x] a task list
- [x] an emoji
- [x] a blockquote
- [x] a hyper link
- [x] inline code
- [x] some text formatting: bold, italic, stikethrough etc.

Commit:
Add Gradle support

Ryan-L98 committed:
Implement Greet, Echo, Exit

Ryan-L98 committed:
Implement Greet, Echo, Exit

Ryan-L98 committed:
Implement Add, List

Ryan-L98 committed:
Implement Mark as Done

Ryan-L98 committed:
Implement ToDos, Events, Deadlines

Ryan-L98 committed:
Automated Text UI Testing

Ryan-L98 committed:
Automated Text UI Testing

Ryan-L98 committed:
Merge branch 'master' of https://github.com/Ryan-L98/ip

Ryan-L98 committed:
Handle Errors

Ryan-L98 committed:
Handle Errors

Ryan-L98 committed:
Implement Delete

Ryan-L98 committed:
Use Enums

Ryan-L98 committed:
Use Enums

Ryan-L98 committed:
Use Enums

Ryan-L98 committed:
Implement Save

Ryan-L98 committed:
Implement Save

Ryan-L98 committed:
Implement Save

Ryan-L98 committed:
Implement Save

Ryan-L98 committed:
Implement Save

Ryan-L98 committed:
Implement Save

Ryan-L98 committed:
Implement Dates and Times

Ryan-L98 committed:
Implement Save

Ryan-L98 committed:
Merge branch 'branch-Level-7'

Ryan-L98 committed:
Merge branch-Level-7 and branch-Level-8

Ryan-L98 committed:
Merge branch-Level-8

Ryan-L98 committed:
Use More OOP

Ryan-L98 committed:
Organize into Packages

Ryan-L98 committed:
Add JUnit Tests

Ryan-L98 committed:
Add JavaDoc comments

Ryan-L98 committed:
Follow the Coding Standard

Ryan-L98 committed:
Implement Find

Ryan-L98 committed:
Commit Test Text Files

Ryan-L98 committed:
Merge branch 'branch-A-CodingStandard'

Ryan-L98 committed:
Merge branch 'branch-Level-9'

Ryan-L98 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Ryan-L98 committed:
Use Gradle

Ryan-L98 committed:
Use CheckStyle

medajet commented on src/main/java/bob/Storage.java:
```diff
@@ -0,0 +1,69 @@
+package bob;
+
+import bob.Task.Task;
+import bob.command.ListCommand;
+import bob.exception.FileException;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.ArrayList;
+import java.util.List;
```
Files are all properly packaged, imports are neatly listed out, same goes for all other classes 👍

medajet commented on src/main/java/bob/Task/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+public class Deadline extends Task {
+    private LocalDateTime date;
+
+    public Deadline(String name, LocalDateTime date) {
+        super(name);
+        super.type = "D";
+        super.setStatus(0);
+        this.date = date;
+    }
+
+    @Override
+    public String printStatus() {
+        return "[D] " + Task.statusSymbols[super.getStatus()] + " " + this + " (by: " +
+                date.format(DateTimeFormatter.ofPattern("dd MMM yyy HH:mm")) + ")";
+    }
+
+    @Override
+    public String toString() {
+        return super.getName();
+    }
```
Consider writing Javdoc comments for public methods if you have time, same goes for the other classes too

Ryan-L98 replied:
Definitely will work on it in the next increments 😅

medajet commented on src/main/java/bob/Ui.java:
```diff
@@ -0,0 +1,100 @@
+import java.util.Scanner;
+
+public class Ui {
+    private final String lineSplit = "====================================================================";
+    private final String logo = "██╗░░██╗██╗  ██╗██╗███╗░░░███╗  ██████╗░░█████╗░██████╗░\n"
```
Constant names must be all uppercase using underscore to separate words, i.e. LINE_SPLIT and LOGO

Ryan-L98 replied:
Thanks for pointing this out! Will make the changes 👍

medajet commented on src/main/java/bob/Task/Task.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Represents the task that the bob program will record.
+ */
+public abstract class Task implements Serializable {
```
Pretty solid idea to declare Task as an abstract class and having the 3 task classes inherit from this👍

Ryan-L98 replied:
Thank you! 😄

medajet reviewed (commented):
Good job on the code! Elaborate use of abstraction and separating different objects into their own classes, and code is very readable as well 👍

Ryan-L98 committed:
Implement GUI

zawedcvg commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,89 @@
+     * @return Command that was interpreted from user input
+     * @throws BobException exceptions that tells Bob to let the user know what went wrong
+     */
+    public static Command parse(String input) throws BobException {
```
Maybe change the name to parseInput/parseUserInput for more readibility

zawedcvg commented on src/main/java/bob/Storage.java:
```diff
@@ -0,0 +1,69 @@
+import bob.exception.FileException;
+
+import java.io.File;
+import java.io.FileInputStream;
```
all the files are properly packaged and imported. Nice job!

zawedcvg commented on src/main/java/bob/Ui.java:
```diff
@@ -0,0 +1,100 @@
+        System.out.println("      (づ｡◕‿‿◕｡)づ");
+    }
+
+    public void say(String input) {
```
The function name is a bit ambiguous here. Maybe switch it to sayInput/displayInput?

zawedcvg commented on src/main/java/bob/command/Command.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Represents the command that will be executed by the program.
+ */
+public abstract class Command {
```
Nice job abstracting the command class

zawedcvg reviewed (commented):
Well written code following the given guidelines in almost all the files. Nice job!

Ryan-L98 committed:
Implement Assertions

Ryan-L98 committed:
Improve code quality

Ryan-L98 committed:
Merge pull request #1 from Ryan-L98/branch-A-Assertions

Implement Assertions

Ryan-L98 committed:
Merge branch 'master' into A-CodeQuality

Ryan-L98 committed:
Merge pull request #2 from Ryan-L98/A-CodeQuality

Improve code quality

Ryan-L98 committed:
Mark and Delete Command: throw InvalidIndexException

Mark and Delete Commands throw InvalidCommandException when the index portion of the command is left blank.

Both commands should throw the InvalidIndexException when index is not provided to give a clearer error message to the user.

Let's update the Parser.parseDelete and Parser.parseMark to throw the InvalidIndexException when applicable.

Ryan-L98 committed:
Implement Priority extension: attach priorities to tasks

Attaching priorities to tasks allows the user to see which tasks need to be done first and which tasks can be delayed.

Let's
* add a priority property to tasks
* add PriorityCommand to set the priority of tasks
* add PriorityListCommand to display tasks based on their priority

Setting priority only after a task is created allows tasks to have no priority which provides more flexibility for the user.

Ryan-L98 committed:
Merge pull request #3 from Ryan-L98/C-Priority

Implement Priority extension: attach priorities to tasks

Ryan-L98 committed:
Implement better GUI

Ryan-L98 committed:
Add representative screenshot

Ryan-L98 committed:
Set theme jekyll-theme-tactile

Ryan-L98 committed:
Add user guide

Ryan-L98 committed:
Merge remote-tracking branch 'origin/master'

Ryan-L98 committed:
Set theme jekyll-theme-slate

Ryan-L98 committed:
Update README.md

Ryan-L98 committed:
Update README.md

Ryan-L98 committed:
Rename and move Ui.png

Ryan-L98 committed:
Move Ui.png into docs
